File: backend/app/schemas/mainline_graph.py
"""主线图谱 API 模型。"""

from datetime import date, datetime
from typing import Any, Literal

from pydantic import BaseModel, Field

from app.schemas.concept_graph import ConceptGraphResponse


class MainlineGraphVersionMeta(BaseModel):
    id: int
    trade_date: date
    kind: str
    title: str | None = None
    status: str
    parent_version_id: int | None = None
    created_by: int | None = None
    published_at: datetime | None = None
    meta: dict[str, Any] = Field(default_factory=dict)
    created_at: datetime | None = None
    updated_at: datetime | None = None


class MainlineGraphNodeItem(BaseModel):
    id: int
    theme_id: int
    theme_name: str = ""
    mainline_score: int = 0
    strength_score: int = 0
    lifecycle_stage: str
    role: str
    payload: dict[str, Any] | None = None


class MainlineGraphEdgeItem(BaseModel):
    id: int
    from_theme_id: int
    to_theme_id: int
    weight: float
    method: str
    status: str
    rationale: str = ""
    created_by: int | None = None


class MainlineGraphViewResponse(BaseModel):
    trade_date: date
    version: MainlineGraphVersionMeta | None = None
    nodes: list[MainlineGraphNodeItem] = Field(default_factory=list)
    edges: list[MainlineGraphEdgeItem] = Field(default_factory=list)
    empty: bool = False


class MainlineGraphVersionListResponse(BaseModel):
    trade_date: date
    items: list[MainlineGraphVersionMeta] = Field(default_factory=list)


class MainlineGraphEnsureRequest(BaseModel):
    trade_date: date | None = None
    use_model: bool = False


class MainlineGraphEnsureResponse(BaseModel):
    trade_date: date
    version_id: int
    node_count: int = 0
    edge_count: int = 0
    model_queued: bool = False
    generated_at: datetime | None = None
    elapsed_ms: int = 0


class MainlineGraphCreateDraftRequest(BaseModel):
    trade_date: date | None = None
    source_version_id: int | None = None
    title: str | None = None


class MainlineGraphEdgePatch(BaseModel):
    op: Literal["upsert", "delete"]
    edge_id: int | None = None
    from_theme_id: int | None = None
    to_theme_id: int | None = None
    weight: float | None = None
    method: str = "manual"
    status: str = "active"
    rationale: str = ""


class MainlineGraphPatchEdgesRequest(BaseModel):
    edges: list[MainlineGraphEdgePatch] = Field(default_factory=list)


class MainlineGraphAcceptEdgeRequest(BaseModel):
    draft_version_id: int


class MainlineGraphThemeConceptResponse(BaseModel):
    theme_id: int
    theme_name: str = ""
    trade_date: date | None = None
    lifecycle_stage: str | None = None
    strength_score: int | None = None
    mainline_score: int | None = None
    concept_graph: ConceptGraphResponse = Field(default_factory=ConceptGraphResponse)

File: frontend/src/types/mainline-graph.ts
/** 主线图谱 API 类型 */

import type { ConceptGraph } from '@/types/theme'

export type MainlineGraphMode = 'narrative' | 'concept'

export type MainlineGraphVersionKind = 'auto' | 'draft' | 'published'

export type MainlineGraphVersionStatus = 'open' | 'published' | 'archived'

export type MainlineGraphEdgeMethod = 'rules' | 'model' | 'manual'

export type MainlineGraphEdgeStatus = 'active' | 'suggested' | 'rejected'

export type MainlineGraphNodeRole = 'mainline' | 'branch' | 'other'

export interface MainlineGraphVersionMeta {
  id: number
  trade_date: string
  kind: string
  title: string | null
  status: string
  parent_version_id: number | null
  created_by: number | null
  published_at: string | null
  meta: Record<string, unknown>
  created_at: string | null
  updated_at: string | null
}

export interface MainlineGraphNodeItem {
  id: number
  theme_id: number
  theme_name: string
  mainline_score: number
  strength_score: number
  lifecycle_stage: string
  role: string
  payload: Record<string, unknown> | null
}

export interface MainlineGraphEdgeItem {
  id: number
  from_theme_id: number
  to_theme_id: number
  weight: number
  method: string
  status: string
  rationale: string
  created_by: number | null
}

export interface MainlineGraphViewResponse {
  trade_date: string
  version: MainlineGraphVersionMeta | null
  nodes: MainlineGraphNodeItem[]
  edges: MainlineGraphEdgeItem[]
  empty: boolean
}

export interface MainlineGraphVersionListResponse {
  trade_date: string
  items: MainlineGraphVersionMeta[]
}

export interface MainlineGraphEnsureResponse {
  trade_date: string
  version_id: number
  node_count: number
  edge_count: number
  model_queued: boolean
  generated_at: string | null
  elapsed_ms: number
}

export interface MainlineGraphThemeConceptResponse {
  theme_id: number
  theme_name: string
  trade_date: string | null
  lifecycle_stage: string | null
  strength_score: number | null
  mainline_score: number | null
  concept_graph: ConceptGraph
}

export interface MainlineGraphViewParams {
  trade_date?: string
  version_id?: number
}

export interface MainlineGraphVersionsParams {
  trade_date?: string
}

export interface MainlineGraphEnsureRequest {
  trade_date?: string
  use_model?: boolean
}

export interface MainlineGraphCreateDraftRequest {
  trade_date?: string
  source_version_id?: number
  title?: string
}

export interface MainlineGraphEdgePatch {
  op: 'upsert' | 'delete'
  edge_id?: number
  from_theme_id?: number
  to_theme_id?: number
  weight?: number
  method?: string
  status?: string
  rationale?: string
}

export interface MainlineGraphPatchEdgesRequest {
  edges: MainlineGraphEdgePatch[]
}

export interface MainlineGraphAcceptEdgeRequest {
  draft_version_id: number
}

export interface MainlineGraphThemeConceptParams {
  trade_date?: string
}
